lib/redis_dedupe/set.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+# :nodoc:
+module RedisDedupe
+  class << self
+    attr_accessor :client
+  end
+
+  # A mechanism to make sure that a block of code will only be called once for each specified identifier (or +member+)
+  # even if the calling process dies or restarts, as long the datastore is +Redis+-backed.
+  #
+  # @example Keep the set around for 7 days, letting Redis handle its own memory cleanup after that time
+  #   ```
+  #   dedupe = RedisDedupe::Set.new($redis, "send_payment_due_emails")
+  #   Account.all do |account|
+  #     dedupe.check(account.id) do
+  #       mail(to: account.billing_email, subject: "PAY US... NOW!!!")
+  #     end
+  #   end
+  #   ```
+  #
+  # @example If you want to be able to repeat the process at any time immediately following this method
+  #   ```
+  #   dedupe = RedisDedupe::Set.new($redis, "send_welcome_emails")
+  #   Account.all.pluck(:email) do |email|
+  #     dedupe.check(email) { mail(to: email, subject: "Hello!") }
+  #   end
+  #   dedupe.finish
+  #   ```
+  #
+  class Set
+    SEVEN_DAYS = 7 * 24 * 60 * 60
+
+    attr_reader :key, :expires_in
+
+    def initialize(redis, key, expires_in = SEVEN_DAYS)
+      @redis      = redis
+      @key        = key
+      @expires_in = expires_in
+    end
+
+    # Ensures that a block of code will only be run if the +member+ is not already contained in Redis.
+    # ie: the code block has not already run for the specified +member+.
+    #
+    # Note that if the given block raises an error, the +member+ will not remain in the +Set+ and may be tried again.
+    #
+    # @param [String, Integer] member identifiying value to make sure the given block only runs once
+    #
+    # @return `nil` if the block was not run, otherwise the result of the yielded block
+    #
+    def check(member)
+      return unless execute_block_for_member?(member)
+
+      # If it didn't get processed... we can retry it again
+      # Unless it it's an error that's ALWAYS going to occur, in which case we might not want to do this.
+      # Should we track/handle this differently?
+      # But as long as the calling code doesn't error in the yielded block, all should be fine.
+      begin
+        yield
+      rescue StandardError => e
+        redis.del(key, member)
```

Comment:
I think the command for member deletion in the set is `srem`? Whereas `del` is for removing the entry for an entire dey. I had to look it up, couldn't recall off-hand if it was `srem` or `sdel`.

https://redis.io/commands/srem

Reply:
oh shoot!  You're right.  Why does that not fail!?!?!  It should give an argument error saying there are 2 args when only 1 is expected!  Oh... maybe it allows you to delete multiple keys?  Yup!  That's it!

OK... adding a fix with a new spec for that. Thanks... great catch!!